### ISSUE TITLE
Add a property to adjust spatial stereo panning on AudioStreamPlayer2D and AudioStreamPlayer3D

### ISSUE THREAD

**Issue by elvisish** [godotengine/godot-proposals#3384]
### Describe the project you are working on

A 3D walking simulator.

### Describe the problem or limitation you are having in your project

I want an ambient sound to get quieter the further I am from it, but I do not want it to pan between the left and right channels depending on my facing angle.

Presently, the distance attenuation can be disabled so the sound appears at the same volume from all distances, but not the spatial position in the stereo field. This makes no sense as both should be able to be enabled or disabled as needed.

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

Using a regular AudioStreamPlayer and having to write a distance attenuation script for it requires parenting the AudioStreamPlayer to a spatial and writing the code for the algorithm required for attenuation which is incredibly inconvenient. This would allow for ambient sounds to play without any changes to the stereo perception, but get quieter or louder from the source.

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

Just a new setting in the inspector, even the ability to adjust the amount of spatial positioning (so rotating wouldn't be so extreme in the left/right channels, even more extreme or none at all).

### If this enhancement will not be used often, can it be worked around with a few lines of script?

It would be AudioStreamPlayer3D code.

### Is there a reason why this should be core and not an add-on in the asset library?

It would be AudioStreamPlayer3D code.

**Comment by Calinou** [comment]
This can be worked around with an AudioStreamPlayer node. You can set the volume depending on the distance relative to a node, and simulate distance attenuation this way.

**Comment by elvisish** [comment]
> This can be worked around with a (non-positional) AudioStreamPlayer node. You can set the volume depending on the distance between the camera (or listener) and another Node3D, and simulate distance attenuation this way. In general, this is more flexible for doing things such as playing area-specific music (since you can also toggle it based on the player colliding with Area3D).
> 
> Could you describe some concrete use cases for playing a spatial sound without panning? I can't see too many of them myself.

it’s not just the the ability to disable it, but also the ability to reduce how much of it there is. At present it’s very jarring, I imagine especially for accessibility users it might be disorientating to have such extreme panning where a more comfortable amount of panning would be preferred.

Writing my own distance code requires implementing writing algorithms that attenuate linearly, quadratically and logarithmically. It also requires writing a unit size method, and Doppler if I wanted to use Doppler without spatial panning.

Most use cases would be for ambience or music, either way it would highly convenient and makes sense as presently to only allow disabling attenuation is illogical and equally as unrealistic as being able to disable spatial panning.

**Comment by Calinou** [comment]
@ellenhp mentioned tweaking SPCAP to allow reducing the panning intensity. This is definitely a desired feature; it's just that we don't have anyone with time to implement it.

**Comment by elvisish** [comment]
> @ellenhp mentioned tweaking SPCAP to allow reducing the panning intensity. This is definitely a desired feature; it's just that we don't have anyone with time to implement it. See also [godotengine/godot#51667](https://github.com/godotengine/godot/pull/51667).

Presumably this would also allow the panning intensity to be effectively removed completely by setting it to 0 (as would give pairity with setting distance attenuation to 0)

**Comment by elvisish** [comment]
I've been trying to come up with a quick alternative for anyone looking, this is closest I've come up with so far:

extends Spatial

onready var audiostream = $AudioStreamPlayer
onready var playercam = get_viewport().get_camera()
var distance

func _ready():
	distance()
	audiostream.play()

func _process(delta):
	distance()

func distance():
	distance = global_transform.origin.distance_squared_to(playercam.global_transform.origin)
	distance *= 0.01
	audiostream.volume_db = -distance

**Comment by elvisish** [comment]
My custom script works fine, but it's not 3D at all and I'm worried about performance. Could a small improvement to AudioStreamPlayer3D that allows for spatial width adjustment be considered for the 3.4 cycle? The issue really is that at present the sound isn't very realistic as it sounds like it's always coming out of one ear and isn't really at all how it would sound in the real world.

**Comment by Calinou** [comment]
> Could a small improvement to AudioStreamPlayer3D that allows for spatial width adjustment be considered for the 3.4 cycle?

[Patch releases no longer get new features](https://godotengine.org/article/versioning-change-godot-3x), so it will have to wait for 3.5 at best.

**Comment by elvisish** [comment]
> [Patch releases no longer get new features](https://godotengine.org/article/versioning-change-godot-3x), so it will have to wait for 3.5 at best.

This was recently added to 3.4.1rc1:
> Input: Add Input.is_physical_key_pressed() method (GH-55251).

Does that not qualify as a new feature (or at least any more so than this would be)?

**Comment by Calinou** [comment]
I discussed this proposal with ellenhp on the Godot Contributors Chat. I think it should work like this:

- Add a **Tightness** property to AudioStreamPlayer2D and AudioStreamPlayer3D to adjust the spatialization intensity on a per-node basis.
- Add a **Tightness Scale** project setting that acts as a *multiplier* for all tightness values. This project setting would be linked to an AudioServer method which can be used to change this setting at run-time. This setting can be used to adjust spatialization on a project-wide basis, which is useful to better support both speakers and headphones setups in the same project. (When using headphones, you need spatialization to be less strong compared to a speakers setup.)

**Comment by elvisish** [comment]
> I discussed this proposal with ellenhp on the Godot Contributors Chat. I think it should work like this:
> 
> * Add a **Tightness** property to AudioStreamPlayer2D and AudioStreamPlayer3D to adjust the spatialization intensity on a per-node basis.
> * Add a **Tightness Scale** project setting that acts as a _multiplier_ for all tightness values. This project setting would be linked to an AudioServer method which can be used to change this setting at run-time. This setting can be used to adjust spatialization on a project-wide basis, which is useful to better support both speakers and headphones setups in the same project. (When using headphones, you need spatialization to be less strong compared to a speakers setup.)

This sounds amazing, thank you! Is it planned for 3.5?

**Comment by Calinou** [comment]
> This sounds amazing, thank you! Is it planned for 3.5?

Only if a motivated contributor does the required work :slightly_smiling_face: 

To get this merged in `3.x`, two things are needed:

- An implementation of this feature on `master` needs to be created from scratch, with the same capabilities as the `3.x` version. The audio server has changed significantly in `master` since https://github.com/godotengine/godot/pull/51296 was merged (it's a backwards-incompatible change).
- https://github.com/godotengine/godot/pull/42358 would need to be rebased against `3.x` and reworked to follow my above comment. Once this is done, another pull request should be opened against `3.x` (as https://github.com/godotengine/godot/pull/42358 is open against `master`, and that can't be changed easily).

**Comment by elvisish** [comment]
Should this be closed? https://github.com/godotengine/godot/pull/58841

**Comment by Calinou** [comment]
> Should this be closed? [godotengine/godot#58841](https://github.com/godotengine/godot/pull/58841)

Yes, as this feature was merged in the `master` branch for 4.0. As I mentioned above, a `3.x` backport requires a full rewrite of the feature, and it won't be tackled by ellenhp or myself.

**Comment by elvisish** [comment]
> > Should this be closed? [godotengine/godot#58841](https://github.com/godotengine/godot/pull/58841)
> 
> Yes, as this feature was merged in the `master` branch for 4.0. As I mentioned above, a `3.x` backport requires a full rewrite of the feature, and it won't be tackled by ellenhp or myself.
> 
> You can keep using the script approach in `3.x`, it should work well enough for now.

Aw that's a shame, I thought it was being done for 3.x, ha! Should there be an open issue for backporting it from 4.0?

**Comment by Calinou** [comment]
> Aw that's a shame, I thought it was being done for 3.x, ha! Should there be an open issue for backporting it from 4.0?

You can open a proposal for this, but it's unlikely to be implemented before 4.0 is released (unless you do it yourself, of course).

**Comment by elvisish** [comment]
https://github.com/godotengine/godot-proposals/issues/4692
